### PR TITLE
Load a save once: tabs stop loading one by one

### PR DESCRIPTION
**Pull request by dh0er**
Opening a savegame and then walking its tabs cost roughly twelve seconds of core
work on a real save (2.5 MB file, 119 MB decoded payload, 1.4M properties), and
every repeat paid the same again. Each tab loaded on first paint, so every first
click sat on a spinner.

Measured on a real save, min of three runs:

| step | before | after |
| --- | --- | --- |
| `inspect_save` | 3883 ms | 2.0 ms¹ |
| world clock | 1073 ms | 2.0 ms |
| character index | 513 ms | 2.0 ms |
| hero attributes | 891 ms | 1.8 ms |
| skills | 700 ms | 1.8 ms |
| NPC list | 257 ms | 1.9 ms |
| quests / glossary / tutorials | 149 / 228 / 154 ms | ~2 ms |
| knowledge / events | 151 / 47 ms | ~2 ms |
| property browser | 603 ms | 2.0 ms |
| backups | 130 ms | 51 ms |
| **total** | **8785 ms** | **78 ms** |

¹ a repeat. The first open of a save is 7.7 s → 1.7 s; that is the part that
genuinely has to be computed.

## What changed

**The typed search allocated three Strings per property.** For each of ~1.4M
properties it built a path segment, a formatted value, and — once per query term
— a lower-cased copy of the whole display path, just to test a substring. The
path now borrows property names out of the tree and owns only the
`[index]`/`{mapKey}` segments it has to format, a lower-cased twin of the
display path is kept in lockstep, and the path and value are built only for a
property that reaches the result page.

**`inspect_save` ran eleven independent read-only passes back to back** — the
FString scan, the typed parse, and one traversal each for the inventory, armor,
slot integrity, progression, NPC, faction, skill and glossary blocks. They share
no state, so they now run on scoped threads in two stages and the load costs the
longest pass instead of their sum. The inventory summary is split so its byte
scan no longer waits for typed-tree traversals it does not depend on.

**`private.skills.list` went around the parsed-root cache** — the one read
command that did. It copied the whole decoded payload out of the byte cache and
re-parsed it.

**No command result was cached.** Read commands now memoize their response under
the request plus a content fingerprint of every file the answer depends on; for
`inspect_save` that includes the sibling `PersistentDataList.sav`, which carries
the slot's profile assignment. `list_backups` and `scan_save_dir` are
deliberately excluded — they describe a folder, which changes without any save
changing. A repeat read costs 2 ms: reading and hashing the file.

**`list_backups` read and hashed every backup file serially.** A folder with a
hundred-odd backups made that a tenth of a second on every save selection; the
candidates are now described in parallel. Disk-bound, so it does not scale with
cores.

**The editor prefetches every tab.** Once an inspection lands, a background
warm-up issues exactly the queries the panels will issue, ordered by how soon
the user can reach them: overview, then Characters, World, and the property
browser. It never touches the loading overlay or reports an error, and it stops
the moment a newer load or a write supersedes it, so it cannot make the user's
own request wait behind the rest of the warm-up. The panels' page sizes moved to
`EditorPageSize` because the cache holds one response per exact request — a
panel that quietly picked its own size would be warmed with an answer it never
asks for.

## Verification

- **Byte-identical output.** Seventeen commands over a frozen copy of a real
  save, including the preview and public-only inspects — 2.2 MB of responses,
  unchanged against the previous core. The typed search was separately diffed
  over eight queries including deep pages and the empty query (3.1 MB).
- New integration tests pin the response cache against serving a stale read
  after a write, after an out-of-band file replacement (a cloud sync, the game
  saving over the slot), and for a directory listing that must not be cached.
- New tests cover the prefetch: what it warms, that its page sizes match the
  panels', that it runs once per inspection, that it never raises the loading
  overlay, and that a newer load stops it.
- 412 Rust tests, 546 Flutter tests, `flutter analyze` clean.

Two tests in `player_events_hero_wiring_test` forbade *every* progression query
while the player had no hero id. Their subject — and their own comment — is the
events query, which is still forbidden and still guarded; the other sections now
legitimately load in the background.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Large changes to core caching, save inspection parallelism, and background prefetch concurrency; correctness depends on fingerprint invalidation and supersede logic, with integration tests added but real-save edge cases (external file replacement, sidecars) matter for stale reads.
> 
> **Overview**
> **Save editor tabs feel instant after open** because the Flutter layer now **background-prefetches** every panel’s core queries as soon as a save finishes loading (`prefetchTabData` on state changes), using shared **`EditorPageSize`** so warm-up requests match real panel pagination. Prefetch **does not** show loading UI, **yields** when a newer load/write supersedes it, walks **multi-page** quest/story/NPC lists, and ends with **`warm_save`** so the decoded tree is ready before NPC detail reads.
> 
> The **Rust core (`gore-save`)** adds a **content-fingerprinted response cache** for read commands (invalidated on write; directory listings excluded), a **`warm_save`** command to re-seed decode/parse caches when `inspect_save` is served from cache, **parallel** `inspect_save` summary passes and **parallel** backup listing reads, **typed property search** allocation reductions, and **`private.skills.list`** now uses the shared parsed-root cache instead of re-parsing.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 917a6d749a48f78e82e06004f5bcdb0d57654961. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 1]
```diff
@@ -607,6 +622,110 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+  ) async {
+    // A newer load (or a write) has taken over: its own prefetch will run, and
+    // continuing here would only make the user's request wait behind ours. A
+    // disposed notifier stops it too — the editor is gone, and touching `state`
+    // after teardown throws.
+    bool superseded() =>
+        !mounted ||
+        seq != _loadSeq ||
+        state.selectedPath != path ||
+        state.isLoading;
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart:715]: Confirmed and fixed in b237794f. `prefetchTabData` now returns without claiming the inspection while `state.isLoading` is set, so the state change that clears the flag brings the page back and starts the warm-up for real. Covered by a regression test that drives the notifier only through state changes, exactly as the page does; it fails without the fix.

File: crates/gore-save/src/lib.rs [hunk 2]
```diff
@@ -5267,6 +5451,124 @@ fn summarize_typed_parse_result(
...
+/// game installation, not the save).
+const CACHEABLE_READ_COMMANDS: &[&str] = &[
+    "inspect_save",
+    "search_typed_properties",
+    "query_progression",
+    "private.skills.list",
+    "private.npc.list",
+    "private.characters.list",
+    "private.npc.attributes",
+    "private.npc.position",
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs:5557]: Confirmed and fixed in b237794f. The placement sidecar is now folded into the cache fingerprint for `private.npc.position`, alongside the existing `PersistentDataList.sav` dependency of `inspect_save`; both are declared in one place (`response_companion_files`). Covered by a test that records and then clears a placement note without touching the save bytes and asserts the answer changes; it fails without the fix.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/lib.rs [hunk 3]
```diff
@@ -367,8 +367,25 @@ impl<'a> Reader<'a> {
 }
 
 pub fn execute_json(input: &str) -> String {
+    // A read command is a pure function of the files it reads, so an identical
+    // request against unchanged files can be answered from the last response.
+    // This is what makes re-opening a save, or returning to a tab, free.
+    let cache_key = read_response_cache_key(input);
+    if let Some(key) = &cache_key {
+        if let Some(hit) = cached_response(key) {
+            return hit;
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs:377]: Confirmed on the store side and fixed in 758a7f3b. The fingerprint is now re-derived after the command and the response is kept only when it still matches, so a stored entry always describes the content its key names. Costs one extra read+hash on a cache miss; a hit is unaffected.

On the serve side I did not add a re-check, deliberately: matching a fingerprint means the files hold byte-identical content to what the entry was built from, so a hit already answers for exactly the bytes just observed. Re-reading before every hit would double the hit cost without changing any answer.

Worth recording that this is not test-covered: the window is between two reads that sit microseconds apart (the key derivation and the command's own `fs::read`), not the command duration. A timing-based test I wrote passed just as readily with the guard removed, so I dropped it rather than ship a test that cannot fail.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 4]
```diff
@@ -607,6 +622,117 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+
+    // Ordered by how soon the user can reach the data: the Overview tab is
+    // already on screen, Characters is one click away, then World, then the
+    // property browser.
+    await step(loadGameTime);
+    // Also settles the hero GlobalId that the player's Events sub-tab needs.
+    await step(loadAllCharacters);
+    await step(loadHeroAttributes);
+    await step(loadSkills);
+    await step(loadAllNpcActors);
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart]: Good catch on the interaction; fixed in 66d1fcd3. The memo itself predates this branch and is deliberate — the whole editor is pinned to one inspection, and every other panel reloads only when a new one lands — but the warm-up had no business filling it. It now pages the roster into the CORE cache only, and the panel's own call populates the memo from the file as of that moment; the paging it repeats is answered from the warm cache, so the visit stays fast. Covered by a test that fails when the warm-up fills the memo.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/properties.rs [hunk 5]
```diff
@@ -658,16 +659,72 @@ pub fn search_properties(
...
+            self.display.push_str(" › ");
+            self.lower.push_str(" › ");
+        }
+        self.display.push_str(&segment);
+        // Lower-casing per segment rather than per leaf is what makes the match
+        // cheap. It agrees with lower-casing the joined path: every segment is
+        // followed by a separator or the end of the string, so no character sits
+        // in a different word-final position in one form than in the other.
+        self.lower
+            .extend(segment.chars().flat_map(char::to_lowercase));
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/properties.rs]: Right, and it disproves the reasoning I had written into the comment there. I had argued that segment boundaries never move a character into a different word position — true, but beside the point, since `char::to_lowercase` has no context to apply at all and maps a word-final Σ to σ regardless. Verified: `"ΟΣ"` becomes `"οσ"` char-wise and `"ος"` string-wise, and the query normalizes to `"ος"`, so the property could not be found by its own name.

Fixed in 72c29dae by lower-casing each segment as a string. ASCII takes a fast path that lower-cases in place, so the per-property allocation this walk exists to avoid stays avoided; only a non-ASCII segment allocates. Covered by a test over a word-final and a word-medial sigma, which fails on the previous mapping, and typed-search output over a real save is unchanged.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 6]
```diff
@@ -607,6 +622,122 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+    if (inspection == null || path == null) return;
+    // The inspection lands BEFORE its load finishes — `_inspect` still has the
+    // backup list to fetch — and every warm-up step bails out while a load is in
+    // flight. Claiming the inspection here would therefore burn it on a warm-up
+    // that does nothing, and the identity check below would refuse to try again.
+    // Wait instead: clearing the loading flag is itself a state change, so the
+    // page calls this again, and that call starts the warm-up for real.
+    if (state.isLoading) return;
+    if (identical(_prefetchedFor, inspection)) return;
+    _prefetchedFor = inspection;
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart]: Confirmed and fixed in 3a2f0233. `refreshBackups` does bump `_loadSeq` without producing a new inspection, so every remaining step skipped and the identity check refused to retry — the tabs those steps covered went back to loading the slow way for the session.

An inspection is now retired only after a run that warmed every step; anything less leaves the marker unset so the next state change starts the sequence over, and the steps that already completed come back from the core cache in milliseconds. Since the marker can no longer double as the in-flight guard — and the warm-up itself changes state, the character index settles the hero id — a separate `_prefetchRunning` flag keeps it to one run at a time. Covered by a test that interrupts a warm-up and asserts the next trigger picks it up, and that a completed one is still retired.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 7]
```diff
@@ -607,6 +622,155 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+    // page calls this again, and that call starts the warm-up for real.
+    if (state.isLoading) return;
+    if (_prefetchRunning) return;
+    if (identical(_prefetchedFor, inspection)) return;
+    _prefetchRunning = true;
+    prefetchInFlight = _prefetchTabData(
+      path,
+      inspection,
+      _loadSeq,
+    ).whenComplete(() => _prefetchRunning = false);
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart]: Right, and it is a hole in the previous commit's fix. Leaving the inspection unmarked only helps if something calls the warm-up again, and a step still in flight holds `_prefetchRunning` past the moment the interrupting operation clears `isLoading` — so that state change bounces off the guard, and by the time the drained run releases the flag there is nothing left to trigger anything.

Fixed in 28d844e9: a run that ends without warming everything re-arms itself. It cannot spin — a fresh run takes the current load sequence, so it can only be cut short by a NEW interruption, and the entry guards still decide whether it may start.

Worth noting my first attempt at the test did not reproduce this: it ended the drained run BEFORE `isLoading` cleared, which the plain state-change path already covers. The test now holds the first warm-up step past the interruption and drives the notifier through the page's own listener, so the restart cannot come from a trigger the test made by hand. It fails without the re-arm.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/lib.rs [hunk 8]
```diff
@@ -367,8 +367,40 @@ impl<'a> Reader<'a> {
...
+            // describe, and every later read of those earlier bytes would be
+            // served the wrong answer for as long as the entry lives. So keep a
+            // response only when its inputs held still for the whole command.
+            //
+            // A cache HIT needs no such re-check: matching the fingerprint means
+            // the files hold byte-identical content to what the entry was built
+            // from, whatever has happened in between.
+            if let Some(key) = cache_key {
+                if read_response_cache_key(input).is_some_and(|current| current == key) {
+                    store_cached_response(key, &response);
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs:412]: Correct — before/after equality does not prove stability, and my comment claimed it did. Fixed the claim in 39914fa5 rather than the check, deliberately.

Closing the ABA properly means handing the command the bytes the key was taken from instead of letting it read for itself. That is a change to how every read command receives its save — two choke points plus the companion files each command reads separately — not a change to this check, and not one I want to make inside a performance branch six review rounds deep.

What is left needs all of: a replacement landing in the moment between two reads that sit microseconds apart, the original bytes returning before the command ends, and BOTH transitions coming from outside the editor — a write or restore performed here calls `invalidate_decoded_payload_cache`, which drops the save's cached responses outright. The reasoning is now recorded where the check lives so the next reader is not misled about the guarantee.

Happy to take the snapshot-threading change as its own PR if you would rather have it closed than documented.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 9]
```diff
@@ -607,6 +622,164 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+      ),
+    );
+    // The player's Events pane keys on the hero id the character index above
+    // settles. Without one there is nothing to warm — and nothing the pane will
+    // ask for either.
+    final heroId = superseded() ? null : state.heroGlobalId;
+    if (heroId != null) {
+      await step(() => loadMemoryEvents(heroId, limit: EditorPageSize.detail));
+    }
+    await step(() => loadProgressionQuests(limit: EditorPageSize.fullList));
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart]: Confirmed and fixed in 6b7291cb. Both loaders are now walked the way the panels walk them, mirroring their offset arithmetic — the core caches one response per exact request, so an offset the panel never asks for warms nothing.

For scale on the save I measured against: 472 quests and 470 story entries with `includeUnset`, so both sit inside the clamp today and the change costs exactly the same single request there. It matters once either outgrows a page, which is presumably why the panels page at all. Covered by a test over a section with three pages and one with exactly one page, which fails when only the first is warmed.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 10]
```diff
@@ -3156,40 +3344,77 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+  ) async {
+    var offset = 0;
+    while (true) {
+      final result = await page(offset);
+      offset += result.count;
+      // An empty page also covers the failure case, where the loader reports a
+      // zero total: never loop on a stuck or erroring section.
+      if (result.count == 0 || offset >= result.total) break;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart:3502]: Both parts confirmed and fixed in 369bd5da. The walk now checks before every page, not just before it begins, and the quest pages are pinned to the file the walk started against — `loadProgressionQuests` grows the `path` parameter its story counterpart already had, defaulting to the current selection so the panels' requests and the cache keys are unchanged. The NPC walk gets the same check; a real load passes none, since a panel that asked for the roster needs all of it rather than a prefix.

My first test for this proved nothing and I rewrote it: disturbing the editor from outside landed the interruption before the walk had even started, so the sequence skipped it wholesale and the request counts matched with and without the check. It now disturbs the editor from INSIDE the walk on its first page — 40 requests without the check against 20 with it.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/properties.rs [hunk 11]
```diff
@@ -723,71 +800,92 @@ fn scalar_display(value: &PropertyValue) -> Option<String> {
...
-        let display_len = display.len();
-        if !display.is_empty() {
-            display.push_str(" › ");
-        }
-        display.push_str(&p.name);
-        path.push(p.name.to_string());
-        let addressable =
-            ancestors_addressable && name_counts.get(p.name.as_str()).copied() == Some(1);
+        let mark = path.push(Cow::Borrowed(p.name.as_str()), true);
+        let addressable = ancestors_addressable && occurs_once(props, &p.name);
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/properties.rs]: Confirmed — and it is my own regression from the first commit in this branch, where I replaced the per-list name map with a scan per property. Right for the handful of entries a property list normally has, quadratic for a list that has many. Measured on one wide list, scanning every time against counting once:

```
  1,000 siblings    2.4 ms  ->  0.15 ms
  5,000 siblings   52.9 ms  ->  0.46 ms
 20,000 siblings  753.1 ms  ->  3.0 ms
 50,000 siblings    6.1 s   ->  7.5 ms
```

Fixed in fc2b9792: scan below a threshold, count once above it. A test pins that both sides agree on which names are unique — a disagreement would either hide editable properties or offer a duplicated name as editable and let a write resolve to the wrong one. Typed-search output over a real save is unchanged.

File: crates/gore-save/src/lib.rs [hunk 12]
```diff
@@ -5267,6 +5478,142 @@ fn summarize_typed_parse_result(
...
+        return;
+    }
+    guard.push(CachedResponseEntry {
+        key,
+        path: save_path,
+        response: response.to_string(),
+    });
+    // Oldest first: within one save every entry is wanted, so evicting by age
+    // drops the save the user has moved away from rather than the current one.
+    let mut bytes: usize = guard.iter().map(|entry| entry.response.len()).sum();
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs]: Confirmed and fixed in fc2b9792. The budget now counts everything an entry keeps alive — request, fingerprint, path and response — rather than the response alone. The editor itself only sends short queries, but the FFI is a public entry point and the request is held verbatim, so the bound should describe the real footprint.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/lib.rs [hunk 13]
```diff
@@ -4743,6 +4821,58 @@ fn extract_script_paths(data: &[u8]) -> Vec<String> {
...
+/// hundreds of threads.
+fn par_map<T, R>(items: Vec<T>, work: impl Fn(T) -> R + Sync) -> Vec<R>
+where
+    T: Send,
+    R: Send,
+{
+    let threads = std::thread::available_parallelism()
+        .map(|value| value.get())
+        .unwrap_or(4)
+        .min(items.len());
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs:4909]: Capped at eight in df30bdcb, with one correction to the premise: this path passes `include_private: false`, so it never decompresses — a worker holds the file as it sits on disk, and a G1R save is a couple of megabytes rather than the 80 MiB in the example. Against the folder I measured (147 backups, largest 2.6 MB) 24 workers meant about 62 MB, not gigabytes.

The bound is still right, and nearly free, because the work is disk-bound and the curve flattens early — serial 130 ms, then 2 workers 99 ms, 4 workers 70 ms, 8 workers 60 ms, 12 workers 62 ms. Past eight there is nothing left to win and every further reader is another whole file held for it. `par_map` now takes the cap from its caller, since the right bound follows from what a worker holds rather than from how many cores are idle.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: crates/gore-save/src/lib.rs [hunk 14]
```diff
@@ -367,8 +367,52 @@ impl<'a> Reader<'a> {
 }
 
 pub fn execute_json(input: &str) -> String {
+    // A read command is a pure function of the files it reads, so an identical
+    // request against unchanged files can be answered from the last response.
+    // This is what makes re-opening a save, or returning to a tab, free.
+    let cache_key = read_response_cache_key(input);
+    if let Some(key) = &cache_key {
+        if let Some(hit) = cached_response(key) {
+            return hit;
```
[the remark this answers is not in the document]
Comment (dh0er) [reply crates/gore-save/src/lib.rs:377]: Confirmed by measurement and fixed in 83cd26e6. Opening A, switching to B and returning to A, the first read that needs the tree took **1.3 s** — the inspection came back in 4 ms while the caches still held B, and the warm-up steps were all cache hits so nothing reseeded them.

Reseeding on the cached inspection would mean paying the decode and parse there, which is what the hit exists to avoid, so instead there is a `warm_save` command: it asks the core to make a save the one it holds, and the warm-up calls it first, ahead of every step that benefits. The same read then costs 4.7 ms and the 1.3 s happens while the user is still on the Overview tab. On a normally loaded save the call is a cache hit — a file read and a hash.

Not response-cached, since a stored "warmed" would skip the seeding that is the point; a test pins that, and another pins that warming moves the reparse off the following read. Both fail without the fix. They are timing tests because the property is timing — every path returns the same answer — so they compare two measurements from the same run rather than a fixed budget.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

File: apps/save-editor/lib/features/editor/domain/editor_notifier.dart [hunk 15]
```diff
@@ -607,6 +622,196 @@ class EditorNotifier extends StateNotifier<EditorState> {
...
+
+    // First, because everything that reads private data depends on it and the
+    // per-NPC panels cannot be warmed one by one — there are far too many.
+    // Loading a save normally leaves the core holding its decoded payload and
+    // parsed tree, but returning to a save opened earlier does not: the
+    // inspection comes back from the core's cache while the tree it holds still
+    // belongs to whichever save was opened in between, and the first NPC the
+    // user clicks would pay the decode and parse. Cheap when the core already
+    // holds this save, which is the usual case.
+    await step(() => _warmPrivateTree(path));
```
[the remark this answers is not in the document]
Comment (dh0er) [reply apps/save-editor/lib/features/editor/domain/editor_notifier.dart]: Right, and it is the trade-off I got backwards. Moved to the end in f7b77e5a.

The reasoning that settles it: the only case where rebuilding the tree costs anything is a returned-to save, and that is precisely the case where every tab query is already a cached answer — so they belong in front of it, where they cost milliseconds. On a normally loaded save the call is a cache hit wherever it sits. The tabs stay instant either way, and the tree is rebuilt behind them, in time for the first NPC panel, which is the only thing that needed it.

The test that pinned the ordering is inverted with it, and fails if the call moves back to the front.

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review

**Comment by dh0er** [comment]
@cursor review

**Comment by dh0er** [comment]
@codex review
